mm2src/adex_cli/src/adex_proc/response_handler.rs

```diff
@@ -0,0 +1,382 @@
+use anyhow::{anyhow, Result};
+use itertools::Itertools;
+use log::{error, info};
+use mm2_rpc::data::legacy::{BalanceResponse, CoinInitResponse, GetEnabledResponse, Mm2RpcResult, MmVersionResponse,
+                            OrderbookResponse, SellBuyResponse, Status};
+use serde_json::Value as Json;
+use std::cell::RefCell;
+use std::fmt::Debug;
+use std::io::Write;
+
+use super::OrderbookConfig;
+use crate::adex_config::AdexConfig;
+use crate::error_anyhow;
+use common::io::{write_safe_io, writeln_safe_io, WriteSafeIO};
+
+pub(crate) trait ResponseHandler {
+    fn print_response(&self, response: Json) -> Result<()>;
+    fn debug_response<T: Debug + 'static>(&self, response: &T) -> Result<()>;
+    fn on_orderbook_response<Cfg: AdexConfig + 'static>(
+        &self,
+        orderbook: OrderbookResponse,
+        config: &Cfg,
+        otderbook_config: OrderbookConfig,
+    ) -> Result<()>;
+    fn on_get_enabled_response(&self, enabled: &Mm2RpcResult<GetEnabledResponse>) -> Result<()>;
+    fn on_version_response(&self, response: &MmVersionResponse) -> Result<()>;
+    fn on_enable_response(&self, response: &CoinInitResponse) -> Result<()>;
+    fn on_balance_response(&self, response: &BalanceResponse) -> Result<()>;
+    fn on_sell_response(&self, response: &Mm2RpcResult<SellBuyResponse>) -> Result<()>;
+    fn on_buy_response(&self, response: &Mm2RpcResult<SellBuyResponse>) -> Result<()>;
+    fn on_stop_response(&self, response: &Mm2RpcResult<Status>) -> Result<()>;
+}
+
+pub(crate) struct ResponseHandlerImpl<'a> {
+    pub writer: RefCell<&'a mut dyn Write>,
+}
+
+impl ResponseHandler for ResponseHandlerImpl<'_> {
+    fn print_response(&self, result: Json) -> Result<()> {
+        let object = result
+            .as_object()
+            .ok_or_else(|| error_anyhow!("Failed to cast result as object"))?;
+
+        object
+            .iter()
+            .map(SimpleCliTable::from_pair)
+            .for_each(|value| writeln_safe_io!(self.writer.borrow_mut(), "{}: {:?}", value.key, value.value));
+        Ok(())
+    }
+
+    fn debug_response<T: Debug + 'static>(&self, response: &T) -> Result<()> {
+        info!("{response:?}");
+        Ok(())
+    }
+
+    fn on_orderbook_response<Cfg: AdexConfig + 'static>(
+        &self,
+        orderbook: OrderbookResponse,
+        config: &Cfg,
+        otderbook_config: OrderbookConfig,
+    ) -> Result<()> {
+        let mut writer = self.writer.borrow_mut();
+
+        let base_vol_head = "Volume: ".to_string() + &orderbook.base;
+        let rel_price_head = "Price: ".to_string() + &orderbook.rel;
```

Comment:
Don't have hard feelings on this, but using `format` would be quite better I think

Reply:
Thank you )
[done](https://github.com/KomodoPlatform/atomicDEX-API/pull/1768/commits/1ddb260d8b727d8db27355544d69240f03a471c1)